### ISSUE TITLE
Bad image output for Flux.2-dev, rocm, quantization and separate prompt encoding sequence

### ISSUE THREAD

**Issue by john-tecplot** [huggingface/diffusers#13772]
### Describe the bug

Image generation using `black-forest-labs/Flux.2-dev` with diffusers using quantization at int8 and separate stages for prompt encoding and transformer inference results in bad random checkerboard image output. This same workflow works find with all other large models I have tried (QwenImage, z-image, Flux.1-dev, stable-diffusion-3-5-large).

System:
 OS: Fedora 
 Kernel: x86_64 Linux 7.0.8-100.fc43.x86_64
 CPU: Intel Xeon Silver 4114 @ 40x 3GHz [46.0°C]
 GPU: AMD Radeon Pro W7900 (radeonsi, navi31, LLVM 21.1.8, DRM 3.64, 7.0.8-100.fc43.x86_64)
 RAM: 321061MiB

Using docker images: `rocm/pytorch`
tags tested:
- latest (as of May 20, 2026)
- rocm7.2.2_ubuntu24.04_py3.12_pytorch_release_2.7.1

Model resulting in bad output:
- black-forest-labs/Flux.2-dev

The reproduction script included runs the prompt encoding and inference with  an int8 quantization, but explicitly separated by unloading everything in between. 

Output image:

<img width="1024" height="1024" alt="Image" src="https://github.com/user-attachments/assets/d43b26af-16e4-4959-b3a2-0533d966788a" />

### Reproduction

```python
import gc

import diffusers
import torch
import transformers

# tested with these docker images (rocm/pytorch):
#   rocm/pytorch:rocm7.2.2_ubuntu24.04_py3.12_pytorch_release_2.7.1
#   rocm/pytorch:latest (as of 2026-05-20)
# Where latest was at pytorch version 2.8.0

# this seems to make no difference on the output or performance
# torch.backends.cuda.enable_mem_efficient_sdp(False)

model = "black-forest-labs/FLUX.2-dev"
outfile = "cool-cat.png"
prompt = "A cat with a banjo"

print("==== Phase 1: text encoder ====")
print("Loading text encoder (quantization config: llm_int8)...")
te_qconfig = transformers.BitsAndBytesConfig(
    load_in_8bit=True,
)
text_encoder = transformers.Mistral3ForConditionalGeneration.from_pretrained(
    model,
    subfolder="text_encoder",
    quantization_config=te_qconfig,
    tie_word_embeddings=False,
    torch_dtype=torch.bfloat16,
)

print("Building prompt-encoder pipeline (with quantization)...")
encoder_pipeline = diffusers.Flux2Pipeline.from_pretrained(
    model,
    text_encoder=text_encoder,
    transformer=None,
    vae=None,
    torch_dtype=torch.bfloat16,
)

encoder_pipeline.to("cuda")

print("Encoding prompt...")
with torch.no_grad():
    prompt_embeds, text_ids = encoder_pipeline.encode_prompt(prompt=prompt)

print("Unloading prompt-encoder pipeline...")
del encoder_pipeline
del text_encoder
gc.collect()
torch.cuda.empty_cache()

print("==== Phase 2: inference ====")
print("Loading transformer (quantization config: llm_int8)...")
tr_qconfig = diffusers.BitsAndBytesConfig(
    load_in_8bit=True,
)
transformer = diffusers.Flux2Transformer2DModel.from_pretrained(
    model,
    subfolder="transformer",
    quantization_config=tr_qconfig,
    torch_dtype=torch.bfloat16,
)

print("Building inference pipeline (with quantization)...")
pipeline = diffusers.Flux2Pipeline.from_pretrained(
    model,
    text_encoder=None,
    tokenizer=None,
    transformer=transformer,
    torch_dtype=torch.bfloat16,
)

pipeline = pipeline.to("cuda")

print("Running inference...")
result = pipeline(prompt_embeds=prompt_embeds)

print(f"Saving image to {outfile}...")
result.images[0].save(outfile)

print("Done.")
```

### Logs

```shell
# python test-flux2-int8.py
==== Phase 1: text encoder ====
Loading text encoder (quantization config: llm_int8)...
Loading weights: 100%|██████████████████████████████████████████████████████████████████████████████████████████████| 585/585 [04:38<00:00,  2.10it/s]
Building prompt-encoder pipeline (with quantization)...
Loading pipeline components...: 100%|███████████████████████████████████████████████████████████████████████████████████| 3/3 [00:02<00:00,  1.44it/s]
Encoding prompt...
[transformers] Kwargs passed to `processor.__call__` have to be in `processor_kwargs` dict, not in `**kwargs`
/opt/venv/lib/python3.12/site-packages/bitsandbytes/autograd/_functions.py:123: UserWarning: MatMul8bitLt: inputs will be cast from torch.bfloat16 to float16 during quantization
  warnings.warn(f"MatMul8bitLt: inputs will be cast from {A.dtype} to float16 during quantization")
Unloading prompt-encoder pipeline...
==== Phase 2: inference ====
Loading transformer (quantization config: llm_int8)...
Loading checkpoint shards: 100%|████████████████████████████████████████████████████████████████████████████████████████| 7/7 [06:17<00:00, 53.96s/it]
Building inference pipeline (with quantization)...
Loading pipeline components...: 100%|███████████████████████████████████████████████████████████████████████████████████| 3/3 [00:01<00:00,  1.51it/s]
Running inference...
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 50/50 [05:16<00:00,  6.33s/it]
Saving image to cool-cat.png...
Done.
```

### System Info

- 🤗 Diffusers version: 0.38.0
- Platform: Linux-7.0.8-100.fc43.x86_64-x86_64-with-glibc2.39
- Running on Google Colab?: No
- Python version: 3.12.3
- PyTorch version (GPU?): 2.8.0+rocm7.0.0.git64359f59 (True)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Huggingface_hub version: 1.15.0
- Transformers version: 5.8.1
- Accelerate version: 1.13.0
- PEFT version: 0.19.1
- Bitsandbytes version: 0.49.2
- Safetensors version: 0.8.0-rc.0
- xFormers version: not installed
- Accelerator: NA

System:
 OS: Fedora 
 Kernel: x86_64 Linux 7.0.8-100.fc43.x86_64
 Shell: zsh 5.9
 Resolution: 10240x2880
 DE: GNOME 49.7
 WM: Mutter
 WM Theme: Adwaita
 GTK Theme: Adwaita [GTK2/3]
 Icon Theme: Adwaita
 Font: Adwaita Sans 11
 CPU: Intel Xeon Silver 4114 @ 40x 3GHz [46.0°C]
 GPU: AMD Radeon Pro W7900 (radeonsi, navi31, LLVM 21.1.8, DRM 3.64, 7.0.8-100.fc43.x86_64)
 RAM: 321061MiB

### Who can help?

This is general use issue about regular inference with a base model.

@sayakpaul @DN6

**Comment by CoderSATTY** [comment]
Hi @sayakpaul @DN6, I'm interested in looking into this! Could you assign it to me?

**Comment by john-tecplot** [comment]
The linked PR #13798 seems to indicate that we shouldn't (can't?) use 8bit quantization and `bfloat` with `Flux.2-dev`. Is that right? Or rather, would it be better to use `float16` instead of `bfloat16` with 8bit quant?

update: I tested the script using `float16` instead of `bfloat16` and it resulted in the same bad output. I'm still not sure about the feasibility of 8bit quants with flux2-dev which is the main question here.

**Comment by CoderSATTY** [comment]
Yep, and your testing confirms a fundamental limitation
Standard 8-bit quantization is currently unfeasible for FLUX.2-dev, and changing the pipeline to `float16` won't bypass it.

To add some technical context to why this fails: the root cause of the precision collapse is an internal `bitsandbytes` MatMul (matrix multiplication) limitation. The standard 8-bit kernels struggle to process FLUX's extreme activation outliers without internally downcasting or introducing severe rounding errors. This compounds across the massive transformer blocks, ultimately corrupting the image generation regardless of whether you use `float16` or `bfloat16` as the base dtype. 

Fortunately, this MatMul issue does not arise when using **4-bit NF4 quantization** (`bnb_4bit_quant_type="nf4"`). The NF4 implementation is structurally optimized to handle standard normal distributions and natively supports `bfloat16` compute types without triggering that destructive precision loss. 

If memory constraints are a concern, skipping 8-bit entirely and using NF4 remains the most stable and recommended path forward for FLUX models.

**Comment by asomoza** [comment]
Hi @john-tecplot you can also use another quantization backend if you really want to use Flux2.dev with 8bit. 

Also with that amount of RAM (if it is not a typo ~336GB) you don't need to use separate pipelines or quantization, you can use group offloading with leaf and cuda streams unless it doesn't work with ROCm.

**Comment by kushdab** [comment]
The checkerboard artifact pattern with `int8` quantization + separate prompt/transformer stages on ROCm is a dtype/device-map mismatch between the text encoder output and the transformer input.

When you split the pipeline into stages, the text encoder runs first and caches `prompt_embeds` as `float32` (ROCm default for unquantized ops), but the int8-quantized transformer expects `float16`. The checkerboard is the transformer receiving wrong-dtype conditioning tensors.

Fix — cast `prompt_embeds` explicitly before the transformer stage:
```python
prompt_embeds, pooled_prompt_embeds, ... = pipe.encode_prompt(...)

# Cast to match quantized transformer's expected dtype:
prompt_embeds = prompt_embeds.to(torch.float16)
pooled_prompt_embeds = pooled_prompt_embeds.to(torch.float16)

image = pipe(
    prompt_embeds=prompt_embeds,
    pooled_prompt_embeds=pooled_prompt_embeds,
    ...
).images[0]
```

Alternatively, pass `torch_dtype=torch.float16` when loading the text encoder to ensure the encoding stage also runs in `float16` from the start.
